projects/hathor-wallet-headless/0004-tx-transaction.md

```diff
@@ -0,0 +1,493 @@
+	- [NOTE-3]: If `true` and address is not from the wallet, then the interpreter should fail short
+
+>[!NOTE]
+>Decision Call: Should we keep the default behavior or `address` or make it more strict? Like, if there isn't any `address` value, then fail short.
+>
+>To counter the behavior of a stricter `address` we can add a flag to let the user choose the desired behavior, either get a new address or get an address by index.
+>
+>Or we can make `address` a poli typed property which can accept `number` for indexes, the string `new` or a general `string` to denote an address.
+
+#### `DataOutputTemplateInstruction`
+
+```ts
+interface DataOutputTemplateInstruction extends InputOutputTemplateInstruction, BaseTemplateInstruction {
+  data: string;
+}
+```
+
+- `data`: UTF-8 encoded string of data
+#### `RawOutputTemplateInstruction`
+
+```ts
+interface RawOutputTemplateInstruction extends InputOutputTemplateInstruction, BaseTemplateInstruction {
+  amount: number;
+  script: string;
+  token?: string;
+  authority?: `mint` | `melt`;
+}
+```
+
+- `amount`: amount of tokens
+- `script`: base64 encoded script
+- `token`: which token to select, defaults to native token, in practice `00` (HTR)
+- `authority`: `mint` or `melt`, if present will create the desired authority output
+	- [NOTE-1]: If the required input is not present, then it will generate an invalid transaction
+
+#### `ShuffleActionTemplateInstruction`
+[shuffleactiontemplateinstruction]: #shuffleactiontemplateinstruction
+
+```ts
+interface RawOutputTemplateInstruction extends ActionTemplateInstruction, BaseTemplateInstruction {
+}
+```
+
+It will apply an array shuffle action over the selected target.
+
+#### `FillChangeActionTemplateInstruction`
+
+```ts
+interface RawOutputTemplateInstruction extends ActionTemplateInstruction, BaseTemplateInstruction {
+}
+```
+
+The `target` doesn't have any effect. It will will check the transaction balance and calculate any change output, if any it will generate a change output and add it to the final position in outputs array.
+
+>[!NOTE]
+>Decision Call: We can add `position` property to allow user determine the position of the change output if it is generated.
+
+### Instruction classes
+
+Each template instruction interface must have its own class implementation. A class instance is useful to both help user to form an instruction and help the interpreter to form the transaction.
+
+- UtxoInputInstruction
+- RawInputInstruction
+- TokenOutputInstruction
+- DataOutputInstruction
+- RawOutputInstruction
+- ShuffleActionInstruction
+- FillChangeActionInstruction
+
+Nevertheless a user can still form an instruction using a literal object.
+
+### Transaction template client
+
+A client gives a user the ability to execute a transaction template object to make a transaction, and gives the user the responsibility to select the wallet, either Wallet Facade of Wallet Service.
+
+```ts
+class TransactionTemplate {
```

Comment:
I don't agree with the `builder` suffix because this conveys the meaning of an specific pattern, which is not the case here, possibly making it a little bit confusing. I rather prefer stick with it or use another nomenclature.